src/documentation/pipelines/test-your-pipeline.md

```diff
@@ -5,4 +5,263 @@ permalink: /documentation/pipelines/test-your-pipeline/
+
+```java  
+static class MyDoFn extends DoFn<String, Integer> { ... }
+MyDoFn myDoFn = ...;
+DoFnTester<String, Integer> fnTester = DoFnTester.of(myDoFn);
+
+TupleTag<String> tag1 = ...;
+TupleTag<Integer> tag2 = ...;
+TupleTagList tags = TupleTagList.of(tag1).and(tag2);
+
+fnTester.setSideOutputTags(tags);
+```
+
+See the `ParDo` documentation on [side inputs](/documentation/programming-guide/#transforms-sideio) for more information.
+
+### Processing Test Inputs and Checking Results
+
+To process the inputs (and thus run the test on your `DoFn`), you call the method `DoFnTester.processBatch`. When you call `processBatch`, you pass one or more main test input values for your `DoFn`. If you set side inputs, the side inputs are available to each batch of main inputs that you provide.
+
+`DoFnTester.processBatch` returns a `List` of outputs—that is, objects of the same type as the `DoFn`'s specified output type. For a `DoFn<String, Integer>`, `processBatch` returns a `List<Integer>`:
+
+```java  
+static class MyDoFn extends DoFn<String, Integer> { ... }
+MyDoFn myDoFn = ...;
+DoFnTester<String, Integer> fnTester = DoFnTester.of(myDoFn);
+
+String testInput = "test1";
+List<Integer> testOutputs = fnTester.processBatch(testInput);
+```
+
+To check the results of `processBatch`, you use JUnit's `Assert.assertThat` method to test if the `List` of outputs contains the values you expect:
+
+```java  
+String testInput = "test1";
+List<Integer> testOutputs = fnTester.processBatch(testInput);
+
+Assert.assertThat(testOutputs, Matchers.hasItems(...));
+
+// Process a larger batch in a single step.
+Assert.assertThat(fnTester.processBatch("input1", "input2", "input3"), Matchers.hasItems(...));
+```
+
+## Testing Composite Transforms
+
+To test a composite transform you've created, you can use the following pattern:
+
+*   Create a `TestPipeline`.
+*   Create some static, known test input data.
+*   Use the `Create` transform to create a `PCollection` of your input data.
+*   `Apply` your composite transform to the input `PCollection` and save the resulting output `PCollection`.
+*   Use `PAssert` and its subclasses to verify that the output `PCollection` contains the elements that you expect.
+
+### Using the SDK Test Classes
+
+[TestPipeline](/https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java) and [PAssert](http://beam.incubator.apache.org/documentation/sdks/javadoc/0.3.0-incubating/org/apache/beam/sdk/testing/PAssert.html) are classes included in the Beam Java SDK specifically for testing transforms.
+
+#### TestPipeline
+
+For tests, use `TestPipeline` in place of `Pipeline` when you create the pipeline object. Unlike `Pipeline.create`, `TestPipeline.create` handles setting `PipelineOptions` interally.
+
+You create a `TestPipeline` as follows:
+
+```java
+Pipeline p = TestPipeline.create();
+```
+
+#### PAssert
```

Comment:
Let's merge in a lot of @tgroh's blog contents?

Ok to leave a link to http://beam.incubator.apache.org/blog/2016/10/20/test-stream.html for now and file an issue to add it later.